# ProLaTherm: *Pro*tein *La*nguage Model-based *Therm*ophilicity Predictor
[![Python 3.8](https://img.shields.io/badge/Python-3.8-3776AB)](https://www.python.org/downloads/release/python-388/)

In this repository, we publish three parts related to protein thermophilicity prediction:

- Code to get predictions using ProLaTherm, the first protein language model-based thermophilicity predictor
- Code to run our protein thermophilicity prediction optimization framework including several feature- and sequence-based prediction models
- A new benchmark dataset for protein thermophilicity prediction consisting of three significantly updated datasets from literature and newly collected data

## Predict using ProLaTherm
Genereller Warnhinweis ressourcen

Docker workflow erklären

pip install workflow erklären - caution: only tested on ubuntu, use venv 

## Protein Thermophilicity Prediction Optimization Framework
Docker installation

Docker workflow mit unseren datasets

how to run own optimizations with new fasta files


## Data
Here, we publish our new benchmark dataset for protein thermophilicity prediction consisting of significantly updated datasets from literature and newly collected data. 
For details on the data collection and preprocessing, see our publication below.

Within the folder `data`, one can find three subfolders:

- `datasets_w_datasplits`: preprocessed and feature-engineered datasets used for our scientific paper (see below) as .csv-files and so-called *index_files* that contain indices for our datasplits to ensure reproducibility. Both files are generated by our optimization pipeline, when providing fasta files containing thermophilic and non-thermophilic proteins
- `fasta_files`: .fasta-files containing our benchmark data split in files containing *thermo* and *non_thermo* species, both as *_fulldata* and *_preprocessed* (CD-Hit with threshold of 40%, removing sequences with lengths below and above 5th and 95th percentile, see our publication for details)
  - Files with prefix *ProtThermoPred_shared_organisms_*: protein sequences of species that occur within the three previously published datasets and our newly collected data
  - Files with prefix *ProtThermoPred_unique_organisms_*: protein sequences of species that only occur in our newly collected data
  - Further .fasta-files with prefix *ProtThermoPred_* (not containing key words *shared_organisms* and *unique_organisms*): protein sequences of our full dataset without a species-specific split
  - Subfolder `update_of_benchmark_datasets`: .fasta-files containing significantly updated datasets from literature
- `meta_files`: .csv-files with meta information such as species and UniProt history for all datasets


## Contributors
This pipeline is developed and maintained by members of the [Bioinformatics lab](https://bit.cs.tum.de) lead by [Prof. Dr. Dominik Grimm](https://bit.cs.tum.de/team/dominik-grimm/):
- [Florian Haselbeck, M.Sc.](https://bit.cs.tum.de/team/florian-haselbeck/)
- [Maura John, M.Sc.](https://bit.cs.tum.de/team/maura-john/)

## Citation
When using parts of this repository, please cite our publication:

**Superior Protein Thermophilicity Prediction With Protein Language Model Embeddings**  
F Haselbeck, M John, Y Zhang, J Pirnay, JP Fuenzalida-Werner, RD Costa, and DG Grimm
*currently under review*

Keywords: Protein Thermophilicity Prediction, Protein Language Model, Machine Learning
